File: advent-of-code/2015/04/01/solution.clj
(ns solution
  (:require [clojure.java.io :as io]
            [clojure.string :as str])
  (:import [java.security MessageDigest]))

(defn find-advent-coin [secret-key]
  (let [md5 (MessageDigest/getInstance "MD5")]
    (loop [n 1]
      (let [text (str secret-key n)
            hash (.digest md5 (.getBytes text "UTF-8"))]
        ;; Check if first 2.5 bytes are zero (5 hex digits)
        ;; hash[0] == 0 -> "00"
        ;; hash[1] == 0 -> "00"
        ;; (hash[2] & 0xF0) == 0 -> "0?"
        (if (and (= 0 (aget hash 0))
                 (= 0 (aget hash 1))
                 (= 0 (bit-and (aget hash 2) 0xF0)))
          n
          (recur (inc n)))))))

;; Test cases with assertions
(assert (= 609043 (find-advent-coin "abcdef")))
(assert (= 1048970 (find-advent-coin "pqrstuv")))

;; Read input and calculate result
(let [input-path (str (.getParent (io/file *file*)) "/../input.txt")
      secret-key (str/trim (slurp input-path))
      result (find-advent-coin secret-key)]
  (println (str "Clojure: " result)))

File: advent-of-code/2015/04/01/solution.go
package main

import (
	"crypto/md5"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
)

func findAdventCoin(secretKey string) int {
	number := 1
	for {
		text := fmt.Sprintf("%s%d", secretKey, number)
		hash := md5.Sum([]byte(text))
		// Check if first 2.5 bytes are zero (5 hex digits)
		if hash[0] == 0 && hash[1] == 0 && (hash[2]&0xF0) == 0 {
			return number
		}
		number++
	}
}

func main() {
	// Test cases with assertions
	if findAdventCoin("abcdef") != 609043 {
		panic("Test failed: findAdventCoin(\"abcdef\") != 609043")
	}
	if findAdventCoin("pqrstuv") != 1048970 {
		panic("Test failed: findAdventCoin(\"pqrstuv\") != 1048970")
	}

	// Read input and calculate result
	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)
	inputPath := filepath.Join(dir, "..", "input.txt")
	content, err := os.ReadFile(inputPath)
	if err != nil {
		panic(err)
	}
	secretKey := strings.TrimSpace(string(content))

	result := findAdventCoin(secretKey)
	fmt.Printf("Go: %d\n", result)
}
